webvtt.go

```diff
@@ -306,45 +305,64 @@ func parseTextWebVTT(i string) (o Line) {
 	// Create tokenizer
 	tr := html.NewTokenizer(strings.NewReader(i))
 
+	webVTTTagStack := make([]*WebVTTTag, 64)
```

Comment:
Could you explain why you're using an intermediate stack with an index here? 🤔

Reply:
As I know, go 1.13 (or even newer) do not have stack, there is proposal but it is never in a release of golang and I don't know why: https://pkg.go.dev/github.com/golang-collections/collections@v0.0.0-20130729185459-604e922904d3/stack. Thus, I choose this way to implement a simple stack. Besides, a line of WebVTT is in XML format (with the exception is `<v>` tag, the `</v>` can be deduced if the voice is applied for the whole line), so for parsing it, we need recursion because XML is a parent-children structure. To make the function simple and avoid the complexity in recursion, stack here for de-recursion

Reply:
I may have misunderstood the logic but that seems overcomplicated 🤔  but I may be wrong 🤔 

What about simplifying to:

```go
tags := []WebVTTTag{}
for {
  ...
  switch t {
    case html.EndTagToken:
      tags = tags[:len(tags)-1]
    case html.StartTagToken:
      // Parse tag
      tags = append(tags, WebVTTTag{
	        Name:       tagName,
	        Classes:    classes,
	        Annotation: annotation,
        })
    case html.TextToken:
      if len(tags) > 0 {
	sa = &StyleAttributes{}
        sa.tags = make([]WebVTTTag, len(tags))
	copy(sa.tags, tags)
	sa.propagateWebVTTAttributes()
      }
  }
}
```

Reply:
This is also a stack, right? Just eliminate the index (the index is now actually the length of slice), I wonder if the performance is better or equal :thinking: . Let me do some benchmark then, if the performance is good enough, this should be a better approach for implementing a simple stack.

Reply:
The behavior is similar to a stack, but I find it more readable which is better for maintainability purposes 👍

Reply:
I did some benchmark and this is the result of new approach:
```
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkIngestStack-8   	   10000	    135770 ns/op	  141616 B/op	     847 allocs/op
PASS
ok  	github.com/edgeware/sbgo/cmd/ew-vodingest	1.386s
```
And the current approach in this PR:
```
cpu: 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz
BenchmarkIngestStack1-8   	   21097	     57454 ns/op	   33408 B/op	     696 allocs/op
PASS
ok  	github.com/edgeware/sbgo/cmd/ew-vodingest	1.806s
```
As you can see, the current approach is two time faster than the new approach, so I am going keep the current approach with some comment about push-pop from the stack.

Benchmark code:
```go
type TestStruct struct {
	field1 string
	field2 string
	field3 []string
}

func BenchmarkIngestStackNew(b *testing.B) {
	testCases := []int{63, 4, 6, 15, 3, 4, 9, 12, 8, 63, 12, 34, 55, 42, 32, 33, 45, 12, 33, 45, 63, 21, 21, 21, 23, 12, 5}
	for i := 0; i < b.N; i++ {
		for _, tc := range testCases {
			var stack []TestStruct
			for j := 0; j < tc; j++ {
				stack = append(stack, TestStruct{"a", "v", []string{"a", "b", "c"}})
			}
			for j := 0; j < tc; j++ {
				stack = stack[:len(stack)-1]
			}
		}
	}
}

func BenchmarkIngestStackOriginal(b *testing.B) {
	testCases := []int{63, 4, 6, 15, 3, 4, 9, 12, 8, 63, 12, 34, 55, 42, 32, 33, 45, 12, 33, 45, 63, 21, 21, 21, 23, 12, 5}
	for i := 0; i < b.N; i++ {
		for _, tc := range testCases {
			stack := make([]TestStruct, 64)
			index := -1
			for j := 0; j < tc; j++ {
				index++
				stack[index] = TestStruct{"a", "v", []string{"a", "b", "c"}}
			}
			for j := 0; j < tc; j++ {
				index--
			}
		}
	}
}
```

Reply:
Could you make a benchmark of the whole `parseTextWebVTT` function instead? I've the feeling the difference you measured is negligible compared to the whole function (e.g. regexp execution). If that's the case, I'd rather have a simpler logic to maintain than a slightly more efficient one. Besides, won't your code panic if there are more than 64 items in the stack?

Reply:
> Besides, won't your code panic if there are more than 64 items in the stack?
Yes, of course, but I think no one will make a line with 64 styling

Reply:
Actually, the new approach is still slower, but it is because of allocation of the stack whenever we push a new item into the stack. So, to mitigate the times we allocate memory, I suggest to merge two approach, like this:
```go
tags := make([]WebVTTTag, 0, 16)
for {
  ...
  switch t {
    case html.EndTagToken:
      tags = tags[:len(tags)-1]
    case html.StartTagToken:
      // Parse tag
      tags = append(tags, WebVTTTag{
	        Name:       tagName,
	        Classes:    classes,
	        Annotation: annotation,
        })
    case html.TextToken:
      if len(tags) > 0 {
	sa = &StyleAttributes{}
        sa.tags = make([]WebVTTTag, len(tags))
	copy(sa.tags, tags)
	sa.propagateWebVTTAttributes()
      }
  }
}
```
16 for initializing the length of stack is enough IMO, no one will put more than 16 styling into a line, even they do that crazy thing, this approach will not crash, just a little bit slower because allocating memory is now triggered again.

Reply:
👍 👍